src/bvals/boundary_conditions_generic.hpp

```diff
@@ -0,0 +1,125 @@
+
+#ifndef BVALS_BOUNDARY_CONDITIONS_GENERIC_HPP_
+#define BVALS_BOUNDARY_CONDITIONS_GENERIC_HPP_
+
+#include <functional>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "basic_types.hpp"
+#include "interface/meshblock_data.hpp"
+#include "interface/sparse_pack.hpp"
+#include "mesh/domain.hpp"
+#include "mesh/meshblock.hpp"
+
+namespace parthenon {
+namespace BoundaryFunction {
+
+enum class BCSide { Inner, Outer };
+enum class BCType { Outflow, Reflect, ConstantDeriv, Fixed };
+
+template <CoordinateDirection DIR, BCSide SIDE, BCType TYPE, class... var_ts>
+void GenericBC(std::shared_ptr<MeshBlockData<Real>> &rc, bool coarse,
+               TopologicalElement el, Real val) {
+  // make sure DIR is X[123]DIR so we don't have to check again
+  static_assert(DIR == X1DIR || DIR == X2DIR || DIR == X3DIR, "DIR must be X[123]DIR");
+
+  // convenient shorthands
+  constexpr bool X1 = (DIR == X1DIR);
+  constexpr bool X2 = (DIR == X2DIR);
+  constexpr bool X3 = (DIR == X3DIR);
+  constexpr bool INNER = (SIDE == BCSide::Inner);
+
+  std::vector<MetadataFlag> flags{Metadata::FillGhost};
+  if (GetTopologicalType(el) == TopologicalType::Cell) flags.push_back(Metadata::Cell);
+  if (GetTopologicalType(el) == TopologicalType::Face) flags.push_back(Metadata::Face);
+  if (GetTopologicalType(el) == TopologicalType::Edge) flags.push_back(Metadata::Edge);
+  if (GetTopologicalType(el) == TopologicalType::Node) flags.push_back(Metadata::Node);
+
+  constexpr bool fluxes = false;
+  auto q = SparsePack<var_ts...>::Get(rc.get(), flags, fluxes, coarse);
+  const int b = 0;
+  const int lstart = q.GetLowerBoundHost(b);
+  const int lend = q.GetUpperBoundHost(b);
+  if (lend < lstart) return;
+  auto nb = IndexRange{lstart, lend};
+
+  std::shared_ptr<MeshBlock> pmb = rc->GetBlockPointer();
+  const auto &bounds = coarse ? pmb->c_cellbounds : pmb->cellbounds;
+
+  const auto &range = X1 ? bounds.GetBoundsI(IndexDomain::interior, el)
+                         : (X2 ? bounds.GetBoundsJ(IndexDomain::interior, el)
+                               : bounds.GetBoundsK(IndexDomain::interior, el));
+  const int ref = INNER ? range.s : range.e;
+
+  std::string label = (TYPE == BCType::Reflect ? "Reflect" : "Outflow");
+  label += (INNER ? "Inner" : "Outer");
+  label += "X" + std::to_string(DIR);
+
+  constexpr IndexDomain domain =
+      INNER ? (X1 ? IndexDomain::inner_x1
+                  : (X2 ? IndexDomain::inner_x2 : IndexDomain::inner_x3))
+            : (X1 ? IndexDomain::outer_x1
+                  : (X2 ? IndexDomain::outer_x2 : IndexDomain::outer_x3));
+
+  // used for reflections
+  const int offset = 2 * ref + (INNER ? -1 : 1);
+
+  // used for derivatives
+  const int offsetin = INNER;
+  const int offsetout = !INNER;
+  pmb->par_for_bndry(
+      label, nb, domain, el, coarse,
+      KOKKOS_LAMBDA(const int &l, const int &k, const int &j, const int &i) {
+        if (TYPE == BCType::Reflect) {
```

Comment:
I tried this but it created all sorts of crazy GPU compiler errors. I think this is because `if constexpr` implies that the code inside the statement doesn't even get compiled if it evaluates to false (I don't have an explanation of why this causes crazy cuda errors, but it seems plausible). I *think* there shouldn't be a substantial performance difference without the `constexpr` since the branch prediction should be very easy.